File: PCA/amplapi/doc/dotnet/_downloads/AsyncExample.cs
using ampl;
using System;
using System.Collections.Generic;
using System.Threading;

namespace Examples
{
  public class AsyncExample
  {
    static List<string> outputs = new List<string>();
    static AutoResetEvent waitHandle = new AutoResetEvent(false);

    public static int Main(string[] args)
    {
      string modelDirectory = ((args != null) && (args.Length > 0)) ? args[0]
             : "../../models";
      string solver = ((args != null) && (args.Length > 1)) ? args[1] : null;

      /*
      // If the AMPL installation directory is not in the system search path:
      ampl.Environment env = new ampl.Environment(
        "full path to the AMPL installation directory");
      // Create an AMPL instance
      using (AMPL a = new AMPL(env)) {}
      */

      using (AMPL ampl = new AMPL())
      {
        if (solver != null) ampl.SetOption("solver", solver);
        ampl.SetOption("reset_initial_guesses", true);
        ampl.SetOption("send_statuses", false);
        ampl.SetOption("relax_integrality", true);

        // Load the AMPL model from file
        ampl.Read(modelDirectory + "/qpmv/qpmv.mod");
        ampl.Read(modelDirectory + "/qpmv/qpmvbit.run");

        // set tables directory (parameter used in the script above)
        ampl.GetParameter("data_dir").Set(modelDirectory + "/qpmv");
        // Read tables
        ampl.ReadTable("assetstable");
        ampl.ReadTable("astrets");

        // set the output handler to accumulate the output messages
        ampl.Output += Ampl_Output;

        // Create the callback object
        Console.WriteLine("Main thread: Model setup complete. Solve on worker thread.");
        // Initiate the async solution process, passing the callback as a parameter. The Action
        // below will be executed by the AMPL API when the solution process will be completed.
        ampl.SolveAsync(() => Callback());

        // Wait for the solution to complete (achieved by waiting on the AutoResetEvent waitHandle
        Console.WriteLine("Main thread: Waiting for solution to end...\n");
        var start = DateTime.Now;
        waitHandle.WaitOne();
        var duration = DateTime.Now - start;

        Console.WriteLine("Main thread: done waiting.");

        // At this stage, the AMPL process is done, the message "Solution process ended." has been
        // printed on the console by the callback and we print a second confirmation from the main thread
        Console.WriteLine("Main thread: waited for {0}", duration.ToString());
        // Print the objective value
        Console.WriteLine("Main thread: cost: {0}", ampl.GetValue("cst").Dbl);
      }
      return 0;
    }

    private static void Ampl_Output(Kind arg1, string arg2)
    {
      outputs.Add(arg2);
    }

    private static void Callback()
    {
      Console.WriteLine("Solution process ended");
      // Signal the waiting thread so that it knows the result is ready.
      waitHandle.Set();
    }
  }
}

File: PCA/amplapi/doc/cpp/_downloads/asyncexample.cc
#include "ampl/ampl.h"

#include <iostream>
#include <thread>
#include <condition_variable>

// Use C++11 synchronization:
std::condition_variable isdone_cv;
std::mutex isdone_mutex;
bool isdone = false;

struct IsDone {
  bool operator()() const { return isdone; }
};

/*
 * Class used as an output handler. It only prints the solver output.
 * Must publicly implement ``ampl::OutputHandler``.
 */
class MyOutputHandler : public ampl::OutputHandler {
public:
  void output(ampl::output::Kind kind, const char* output) {
    if (kind == ampl::output::SOLVE)
      std::printf("Solver: %s\n", output);
  }
};

/*
 * Object used to communicate the end of the async operation. Must
 * publicly implement ``ampl::Runnable``.
 */
class MyInterpretIsOver : public ampl::Runnable {
public:
  void run() {
    std::cout << "Solution process ended. Notifying waiting thread.\n";
    {
      std::lock_guard<std::mutex> lk(isdone_mutex);
      isdone = true;
    }
    isdone_cv.notify_all();
  }
};

int main(int argc, char **argv) {
  try {
    std::string modelDirectory(argc == 3 ? argv[2] : "../models");

    // Create an AMPL instance
    ampl::AMPL ampl;

    /*
    // If the AMPL installation directory is not in the system search path:
    ampl::Environment env("full path to the AMPL installation directory");
    ampl::AMPL ampl(env);
    */

    ampl.setBoolOption("reset_initial_guesses", true);
    ampl.setBoolOption("send_statuses", false);
    ampl.setBoolOption("relax_integrality", true);

    if (argc > 1)
      ampl.setOption("solver", argv[1]);

    // Load the AMPL model from file
    ampl.read(modelDirectory + "/qpmv/qpmv.mod");
    ampl.read(modelDirectory + "/qpmv/qpmvbit.run");

    // Set tables directory (parameter used in the script above)
    ampl.getParameter("data_dir").set(modelDirectory + "/qpmv");
    // Read tables
    ampl.readTable("assetstable");
    ampl.readTable("astrets");

    // Set the output handler to accumulate the output messages
    MyOutputHandler outputHandler;
    ampl.setOutputHandler(&outputHandler);

    // Create the callback object
    MyInterpretIsOver callback;
    std::cout << "Main thread: Model setup complete. Solve on worker thread.\n";
    // Initiate the async solution process, passing the callback object
    // as a parameter.
    // The function run() will be called by the AMPL API when the
    // solution process will be completed.
    ampl.solveAsync(&callback);

    // Wait for the solution to complete (achieved by waiting on the
    // std::condition_variable isdone
    std::cout << "Main thread: Waiting for solution to end...\n\n";
    std::unique_lock<std::mutex> lk(isdone_mutex);

    namespace chrono = std::chrono;
    auto start = chrono::system_clock::now();
    isdone_cv.wait(lk, IsDone());
    auto duration = chrono::system_clock::now() - start;

    std::cout << "Main thread: done waiting.\n";

    // At this stage, the AMPL process is done, the message
    // "Solution process ended." has been printed on the console by the
    // callback and we print a second confirmation from the main thread
    std::cout << "Main thread: waited for "
      << chrono::duration_cast<chrono::milliseconds>(duration).count()
      << " ms\n";
    // Print the objective value
    std::cout << "Main thread: cost: " << ampl.getValue("cst").dbl() << "\n";
  } catch (const std::exception &e) {
    std::cout << e.what() << "\n";
  }
}
